### ISSUE TITLE
impl IntoFuture for IntoIterator<IntoFuture>?

### ISSUE THREAD

**Issue by nyarly** [alexcrichton/futures-rs#662]
By analogy to the JoinN implementations, does it make sense for IntoIterator<IntoFuture> to be IntoFuture via join_all?

**Comment by alexcrichton** [comment]
An interesting idea! I don't think this can be done, however, due to coherence :(

**Comment by nyarly** [comment]
I'm not sure what "coherence" means in this context. Can you direct me to a reference?

**Comment by nyarly** [comment]
Updated: Ilya directed me to 
https://aturon.github.io/blog/2017/02/06/specialization-and-coherence/

**Comment by nyarly** [comment]
Would an implementation for Vec<IntoFuture> or &[IntoFuture] work, then?

**Comment by stuhood** [comment]
As an alternative, might it be possible to allow for `iter().collect::<JoinAll<_>>` by adding `impl FromIterator for JoinAll`?

**Comment by stuhood** [comment]
I looked into this, and was initially optimistic, because the following passes all of the tests:
```
+impl<I> FromIterator<I::Item> for JoinAll<I>
+    where I: IntoIterator,
+          I::Item: IntoFuture,
+{
+    fn from_iter<X: IntoIterator<Item=I::Item>>(iter: X) -> Self {
+        let elems = iter.into_iter().map(|f| {
+            ElemState::Pending(f.into_future())
+        }).collect();
+        JoinAll { elems: elems }
+    }
+}
+
 pub fn join_all<I>(i: I) -> JoinAll<I>
     where I: IntoIterator,
           I::Item: IntoFuture,
 {
-    let elems = i.into_iter().map(|f| {
-        ElemState::Pending(f.into_future())
-    }).collect();
-    JoinAll { elems: elems }
+    i.into_iter().collect()
 }
```
...ie, `join_all` can easily be implemented in terms of `FromIterator for JoinAll`.

But in practice, actually using `into_iter().collect::<JoinAll<_>>()` is challenging, because inference fails in seemingly simple cases, such as:
```
// Fails with "error[E0283]: type annotations required: cannot resolve `_: std::cmp::Eq`": 
assert_done(|| vec![f_ok(1), f_ok(2)].into_iter().collect::<JoinAll<_>>(), Ok(vec![1, 2]));
// Succeeds:
assert_done(|| join_all(vec![f_ok(1), f_ok(2)]), Ok(vec![1, 2]));
```

----

Part of the complexity of the inference would seem to stem from the fact that `JoinAll` places unnecessary bounds on its type parameters. The body of the definition of `JoinAll` does not actually refer to `IntoIterator` at all:
```
pub struct JoinAll<I>
    where I: IntoIterator,
          I::Item: IntoFuture,
{
    elems: Vec<ElemState<<I::Item as IntoFuture>::Future>>,
}
```
...so it would seem that it could be simplified to:
```
pub struct JoinAll<I>
    where I: IntoFuture,
{
    elems: Vec<ElemState<<I as IntoFuture>::Future>>,
}
```

In that case, the inference for `into_iter().collect::<JoinAll<_>>` would be much simpler, because the underscore would represent the iterated type, rather than the `IntoIterator` type.

Certainly a "major version bump" type change, but maybe interesting to explore?

**Comment by stuhood** [comment]
Mm... I guess my previous comment re-discovers #285. After re-implementing/applying #286, I was able to confirm that:
```
assert_done(|| vec![f_ok(1), f_ok(2)].into_iter().collect::<JoinAll<_>>(), Ok(vec![1, 2]));
```
is inferred correctly.

So perhaps once #286 lands this can be re-opened? I've found that using combinators is more legible than using static methods, as static methods require indentation rather than chaining.

**Comment by alexcrichton** [comment]
@nyarly ah yes it is indeed possible to do this for concrete types like `Vec` and slices, but I actually think @stuhood's idea is a great one and could fit very nicely here!

It's a bit of a bummer though that this may be blocked on https://github.com/alexcrichton/futures-rs/pull/286, but thanks for the investigation @stuhood!

**Comment by stuhood** [comment]
@alexcrichton : Thanks! Worth re-titling and re-opening, or shall we open a new issue with regard to `impl FromIterator for JoinAll`?

**Comment by alexcrichton** [comment]
Makes sense to me to retitle, @nyarly do you agree that such an impl would solve your original use case?

**Comment by nyarly** [comment]
Honestly, I'm not sure. I'm looking at:

```rust
// I'm eliding a bunch of "and_then" into a pretend url2future! macro
future::join_all(urls.iter().map(|url| url2future!(url)).collect::<Vec<_>>())
```

I'm not sure if
```rust
urls.iter().map(|url| url2future!(url)).collect::<JoinAll<_>>()
```
is preferable?

If `Vec<IntoFuture>` were `InfoFuture` I could
```rust
urls.iter().map(|url| url2future!(url))
```

Is there another benefit to `FromIterator for JoinAll`?

**Comment by stuhood** [comment]
@nyarly : If you're executing in a position that is expecting an `(Into)Future`, it's possible (unconfirmed) that inference would make the explicit parameter to collect unnecessary: so it would be (imaginary left hand side):
```
let f: impl Future<_> = urls.iter().map(|url| url2future!(url)).collect();
// vs
let f: impl Future<_> = urls.iter().map(|url| url2future!(url));
```
